### PR TITLE
Recover from lifetimes with default lifetimes in generic args

### LINKED ISSUE #107492: Attempt to set a default value for a lifetime generic parameter produces confusing diagnostic
### Code

```Rust
pub struct DefaultLifetime<'a, 'b = 'static> {
    _marker: std::marker::PhantomData<&'a &'b ()>,
}
```

### Current output

```Shell
error: expected one of `,`, `:`, or `>`, found `=`
  --> src/lib.rs:20:39
   |
20 |     pub struct DefaultLifetime<'a, 'b = 'static> {
   |                                    -- ^ expected one of `,`, `:`, or `>`
   |                                    |
   |                                    maybe try to close unmatched angle bracket
```

### Desired output

```Shell
error: expected one of `,`, `:`, or `>`, found `=`
  --> src/lib.rs:20:39
   |
20 |     pub struct DefaultLifetime<'a, 'b = 'static> {
   |                                       ^^^^^^^^^ lifetime parameters cannot have default values
```

### Rationale and extra context

AFAIK, lifetime generic parameters are not currently allowed to have default values. Attempting to use the "obvious" syntax for setting a default lifetime is an opportunity to teach the user that this is not a supported feature, and certainly shouldn't claim that there's an unmatched angle bracket just because it didn't expect the `=` during parsing.

This is happening on `rustc 1.69.0-nightly (d7948c843 2023-01-26)` with the 2021 edition.

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by lenko-d**
Fixes [#107492](https://github.com/rust-lang/rust/issues/107492)

### CODE REVIEW

File: compiler/rustc_parse/src/parser/generics.rs [hunk 1]
```diff
@@ -145,6 +145,19 @@ impl<'a> Parser<'a> {
                         } else {
                             (None, Vec::new())
                         };
+
+                        if this.eat(&token::Eq) {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/generics.rs:149]: This should probably use `check` to check for the `=`, and `lookahead(1, ..)` to check for the lifetime. This is to make sure we don't eat `=` in the case we have `<'a =` without a lifetime that follows.

File: compiler/rustc_parse/src/parser/generics.rs [hunk 2]
```diff
@@ -145,6 +145,19 @@ impl<'a> Parser<'a> {
...
+                            if this.check_lifetime() {
+                                let _lifetime = this.expect_lifetime();
+                                this.struct_span_err(
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/generics.rs:152]: Can you migrate this to a `derive(Diagnostic)`? https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-structs.html

**Review by compiler-errors — COMMENTED** [review]
The general approach looks fine, but I have some changes.

Can you give this PR a more descriptive name? Instead of explaining the problem, it should explain the solution... maybe "recover from lifetimes with default lifetimes in generic args", or something else.

File: compiler/rustc_parse/src/parser/generics.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/generics.rs:149]: The `check_noexpect` should also fix the CI failures

File: compiler/rustc_parse/src/parser/generics.rs [hunk 3]
```diff
@@ -146,16 +147,14 @@ impl<'a> Parser<'a> {
...
-                                this.struct_span_err(
-                                    this.prev_token.span,
-                                    "unexpected default lifetime parameter",
-                                )
-                                .span_label(this.prev_token.span, "lifetime parameters cannot have default values")
-                                .emit();
-                           }
+                        if this.check_noexpect(&token::Eq) {
+                            if this.look_ahead(1, |t| t.is_lifetime()) {
+                                this.sess.emit_err(
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/generics.rs:152]: Sorry, I wasn't super clear when I gave my previous suggestion. We should still consume the default lifetime here:

```suggestion
                        if this.check_noexpect(&token::Eq) && this.look_ahead(1, |t| t.is_lifetime()) {
                                self.bump(2); // Consume `=` and `'lifetime`
                                this.sess.emit_err(
```

File: tests/ui/lifetimes/issue-107492-default-value-for-lifetime.rs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 pub struct DefaultLifetime<'a, 'b = 'static> {
                                    //~^ ERROR unexpected default lifetime parameter
+                                   //~^^ ERROR expected one of `,`, `:`, or `>`, found `=`
```
Comment (compiler-errors) [inline tests/ui/lifetimes/issue-107492-default-value-for-lifetime.rs:3]:
```suggestion
                                   //~| ERROR expected one of `,`, `:`, or `>`, found `=`
```

nit: this chaining style is preferred, but once you apply the above suggestion you probably will only have one error.

File: compiler/rustc_parse/src/parser/generics.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/generics.rs:152]: Also, let's adjust the span here a bit:

Before bumping, let's get the span of the current token:

```
let lo = self.token.span;
self.bump(2);
let span = lo.to(self.prev_token.span);
```

Then use `span` in your error message. That should make the error message render like:

```
LL | pub struct DefaultLifetime<'a, 'b = 'static> {
   |                                   ^^^^^^^^^ lifetime parameters cannot have default values
```

I hope that makes sense.

**Review by compiler-errors — COMMENTED** [review]
Thanks, one last change then I think this is ready to approve. Maybe we could also suggest removing the default lifetime param, but I don't think it's super necessary.

**Comment by compiler-errors** [comment]
Also, please rebase off of the merge commit 338aa5cfd4e263303de24e7419f39120f1f67d9e, the repository has a "no merge commit" policy.

**Comment by oli-obk** [comment]
I think your rebase hit the case described in https://rustc-dev-guide.rust-lang.org/git.html?highlight=rebase#git-is-trying-to-rebase-commits-i-didnt-write

This should resolve itself automatically by rebasing over the latest rust-lang/rust master commit again

**Review by compiler-errors — APPROVED** [review]
Cool!

**Pull request merged**